en/05_Contributing/05_Coding_Conventions/01_PHP_Coding_Conventions.md

```diff
@@ -136,6 +136,133 @@ class MyClass extends Class
+- Note that there will be some inconsistencies with legacy code as get accessors have been a combination of "with get", "without get", and "without get with the first letter uppercased"
+
+### File suffixes and prefixes
+
+- This is for new classes and interfaces that are created. Legacy Silvesrtripe code is inconsistent whether or not it follows these conventions.
+- Suffix the classnames with the type of class that is subclassed. The suffixed is sometimes a shortened version of the parent class because it reads better while retaining easy comprehension. Here are some common examples with the parents classes in brackets
+  - `Exception` (Exception)
+  - `Controller` (Controller)
+  - `Extension` (Extension, DataExtension)
+  - `Admin` (ModelAdmin)
+  - `Handler` (RequestHandler)
+  - `Filter` (SearchFilter)
+  - `Form` (Form)
+  - `Field` (FormField)
+  - `Job` (QueuedJob)
+  - `Task` (BuildTask)
+  - `Service` (generic class type that has no parent class)
+Some parent classes have unorthodox naming, in which case suffix with something more standard
+  - `Page` (SiteTree)
+  - `Controller` (LeftAndMain, CMSMain)
+- Do not suffix classes if they are a subclass any of the following:
+  - `DataObject`
+- Suffix interfaces with `Interface`
+- Suffix traits with `Trait`
+- Prefix abstract classes with `Abstract`
+
+### Other naming conventions
+
+- Extensions should match the name of the class they are extending. For example, an extension class that's added to LeftAndMain should be called LeftAndMainExtension
+- When choosing names for variables and methods use the guiding principle "make it as obvious as possible"
+- Ensure that variable names and method naming is consistent. For example, if you are creating a new class and have a variable named `$itemValue`` which is passed to another method in the same class, the paramater name for that method should also be `$itemValue``.
+
+## Directory naming - src dir
+
+- Put similar types of classes in the same directory. For example, put Controllers in a Controllers directory
+- Put DataObjects in a directory called Models
+- Do not group different class types together in a directory relating to a single feature
+- Use initial caps pluralised for the directory naming e.g. use `Controllers` rather than `Controller` or `controllers`
+
+## Properties and accessors
+
+- Do not use public properites, instead use a combination of private properites and public getters and setters
+- For setters use a combination of a `:static` return type and `return $this` to allow for fluent programming
+- Do not use [dynamic properties](https://www.php.net/manual/en/language.oop5.properties.php#language.oop5.properties.dynamic-properties)
+
+## Method and property visibility
+
+- Use private over protected as this reduces the maintained API surface. Only use protected if there's an immediate reason to upgrade from private.
```

Comment:
Are we opening this can of worms discussion again? I disagree with this. I'd go for something more conservative such as

```suggestion
- Constants should be public if they are intended to be used outside of the class hierarchy, protected if they are intended to be used within the class hierarchy, and private if they are only intended to be used inside this one class.
- Prefer private over protected for all new properties. Subclasses can access the properties through the public getter and setter methods.
- Prefer private over protected for new methods, unless there is a clear use case for customising the specific functionality of that method on its own and the method implementation is unlikely to change in the near future.
```

Reply:
Protected adds API.  We have too much API.  

Extensibility can be achieved via extension hooks.  It's a more conscious approach then spamming out protected methods everywhere "just in case" someone wants it

We've been defaulting to private for years now and there is very little in the way of negative feedback from the wider community.

Reply:
> We've been defaulting to private for years now

I haven't been... I've been following the guidelines I've outlined in my suggested edit.

Reply:
Ah yup your edits are basically want I meant just much better communicated

When I say "default to private" I mean "start with private and if there's some reason to use protected then use that".  That said I personally add very little in the way of protected API.

Older Silverstripe code seems like it was very much "start with protected" - [random example](https://github.com/silverstripe/silverstripe-framework/blob/5/src/Forms/DatetimeField.php).  And yes I get the reason why you'd make everything protected (easier for projects to subclass) - but it also makes maintenance more burdensome because because now everything is set in stone as we need to maintain backwards compatibility.

Reply:
This has been discussed to death, and the culmination was this very appropriate summary by Sam: https://github.com/silverstripe/silverstripe-framework/issues/8996#issuecomment-534785681

I'd prefer we stick to that advice.

Reply:
I don't see any difference in meaning or intention between that advice and my suggested edit above.

Reply:
I think the wording of what you have implies more of a "default state" where-as what Sam wrote is more of a "think about it". Specifically the `Prefer private over protected for new methods` bit.

Reply:
Do you have a suggestion for how to better word it within the context of this documentation page?

Reply:
My suggestion is to leave this section as is. It was fine.

Reply:
Sam's recommendation seems aligned with what's being recommended here, e.g. `Don't simply dump class internals in as protected members in the name of extensibility – doing so will be likely to break developers' projects in minor releases`

Possibly my phrasing was just too blunt and has been interpreted as "don't use protected", that's not what my intention is, I'm also saying "think about it", though you should "start with" (default-to) private and if there's an actual reason for going with protected then cool make it protected.

What I want to avoid is what happened in lots of the older files where people just made all the class internals protected and didn't use private at all.

We do need to mention this in this document since this is the place we'll point to when discussing peer review. We don't want to put to a big long RFC whenever this comes up.

Reply:
Perhaps that last part should be part of the documentation.

Reply:
Have updated